File: variants/ruby/lib/protocol.rb
# frozen_string_literal: true

require 'json'
require 'socket'
require 'base64'

# =============================================================================
# protocol.rb — Camada de transporte da variante Ruby.
#
# Adaptação: no projeto original a comunicação usa gRPC (unário no plano de
# controle, streaming no plano de dados). Aqui usamos um protocolo simples,
# porém equivalente em papel: uma requisição = um objeto JSON numa linha, e uma
# resposta = um objeto JSON numa linha, sobre TCP. Bytes de chunk viajam como
# string base64 dentro do JSON. Isso preserva a natureza cliente/servidor e a
# separação controle/dados sem a dependência de gRPC.
# =============================================================================

module DFS
  module Protocol
    module_function

    # Envia uma requisição a host:port e devolve a resposta (Hash) decodificada.
    # Abre a conexão, escreve uma linha JSON, lê uma linha JSON, fecha.
    def request(host, port, payload, timeout = 30)
      sock = Socket.tcp(host, port, connect_timeout: timeout)
      sock.write(JSON.generate(payload) + "\n")
      line = read_line(sock, timeout)
      raise IOError, "conexão fechada por #{host}:#{port}" if line.nil?

      JSON.parse(line)
    ensure
      sock&.close
    end

    # Lê exatamente uma linha (terminada por \n) do socket.
    def read_line(sock, timeout = 30)
      buffer = +''
      loop do
        chunk = sock.read_nonblock(65_536, exception: false)
        case chunk
        when :wait_readable
          ready = IO.select([sock], nil, nil, timeout)
          return nil if ready.nil? # timeout
        when nil
          return nil # EOF
        else
          buffer << chunk
          idx = buffer.index("\n")
          return buffer[0...idx] if idx
        end
      end
    rescue EOFError
      nil
    end

    # Serve conexões TCP numa porta. Para cada conexão lê UMA requisição JSON,
    # chama o bloco handler e devolve a resposta JSON. Cada conexão em sua thread.
    def serve(host, port)
      server = TCPServer.new(host, port)
      loop do
        client = server.accept
        Thread.new(client) do |sock|
          begin
            line = read_line(sock, 60)
            next if line.nil?

            req = JSON.parse(line)
            resp = yield(req) # handler
            sock.write(JSON.generate(resp) + "\n")
          rescue StandardError => e
            begin
              sock.write(JSON.generate('ok' => false, 'error' => e.message) + "\n")
            rescue StandardError
              # cliente já foi embora
            end
          ensure
            sock.close
          end
        end
      end
    end

    def encode_bytes(bytes)
      Base64.strict_encode64(bytes)
    end

    def decode_bytes(str)
      Base64.strict_decode64(str)
    end
  end
end

File: variants/dotnet-csharp/Protocol.cs
// =============================================================================
// Protocol.cs — Camada de transporte da variante .NET/C#.
//
// Adaptação: o original usa gRPC. Aqui, uma requisição = um objeto JSON numa
// linha e uma resposta = um objeto JSON numa linha, sobre TCP. Bytes de chunk
// viajam como string base64 no JSON. Preserva a natureza cliente/servidor e a
// separação controle/dados sem a dependência de gRPC.
// =============================================================================

using System.Net;
using System.Net.Sockets;
using System.Text;
using System.Text.Json;
using System.Text.Json.Nodes;

namespace Dfs;

public static class Protocol
{
    private static readonly UTF8Encoding Utf8 = new(false);

    // Envia uma requisição a host:port e devolve a resposta (JsonObject).
    public static JsonObject Request(string host, int port, JsonObject payload, int timeoutMs = 30000)
    {
        using var client = new TcpClient();
        var connect = client.ConnectAsync(host, port);
        if (!connect.Wait(timeoutMs))
            throw new TimeoutException($"timeout ao conectar em {host}:{port}");

        using var stream = client.GetStream();
        var bytes = Utf8.GetBytes(payload.ToJsonString() + "\n");
        stream.Write(bytes, 0, bytes.Length);
        stream.Flush();

        var line = ReadLine(stream, timeoutMs);
        if (line == null) throw new IOException($"conexão fechada por {host}:{port}");
        return (JsonObject)JsonNode.Parse(line)!;
    }

    // Serve conexões TCP: cada conexão lê UMA requisição JSON, chama o handler e
    // devolve a resposta JSON. Uma thread por conexão.
    public static void Serve(string host, int port, Func<JsonObject, JsonObject> handler)
    {
        var listener = new TcpListener(IPAddress.Parse(host), port);
        listener.Start();
        while (true)
        {
            var client = listener.AcceptTcpClient();
            var t = new Thread(() => HandleConnection(client, handler)) { IsBackground = true };
            t.Start();
        }
    }

    private static void HandleConnection(TcpClient client, Func<JsonObject, JsonObject> handler)
    {
        try
        {
            using (client)
            using (var stream = client.GetStream())
            {
                var line = ReadLine(stream, 60000);
                if (line == null) return;

                JsonObject resp;
                try
                {
                    var req = (JsonObject)JsonNode.Parse(line)!;
                    resp = handler(req);
                }
                catch (Exception e)
                {
                    resp = new JsonObject { ["ok"] = false, ["error"] = e.Message };
                }

                var bytes = Utf8.GetBytes(resp.ToJsonString() + "\n");
                stream.Write(bytes, 0, bytes.Length);
                stream.Flush();
            }
        }
        catch
        {
            // cliente desconectou; ignore
        }
    }

    // Lê exatamente uma linha (terminada por \n) de um stream de rede.
    private static string? ReadLine(NetworkStream stream, int timeoutMs)
    {
        stream.ReadTimeout = timeoutMs;
        var sb = new StringBuilder();
        var buf = new byte[65536];
        var pending = new List<byte>();
        while (true)
        {
            int read;
            try { read = stream.Read(buf, 0, buf.Length); }
            catch (IOException) { return null; } // timeout
            if (read == 0) return pending.Count > 0 ? Utf8.GetString(pending.ToArray()) : null;

            for (int i = 0; i < read; i++)
            {
                if (buf[i] == (byte)'\n')
                    return Utf8.GetString(pending.ToArray());
                pending.Add(buf[i]);
            }
        }
    }

    public static string EncodeBytes(byte[] data) => Convert.ToBase64String(data);
    public static byte[] DecodeBytes(string s) => Convert.FromBase64String(s);

    // ---- Helpers de acesso ao JSON ------------------------------------------
    public static string Str(JsonObject o, string key) => o[key]?.GetValue<string>() ?? "";
    public static int Int(JsonObject o, string key) => o[key]?.GetValue<int>() ?? 0;
    public static long Long(JsonObject o, string key) => o[key]?.GetValue<long>() ?? 0;
    public static bool Bool(JsonObject o, string key) => o[key]?.GetValue<bool>() ?? false;
}
